Add enif_whereis() function

tburghart opened the pull request:
#### Why do we need this new feature?

There are cases when a NIF needs to send a message, using `enif_send()`, to a long-lived process with a registered name.

A common use-case is logging, where asynchronous fire-and-forget messages are the norm.

There can also be cases where a yielding or dirty NIF or background thread may request a callback from a service with additional information it needs to complete its operation, yielding or waiting (with suitable timeouts, etc) until its state has been updated through the NIF module's API.

NIFs can only send messages to pids, and the lack of name resolution leaves a complicated dance between separate monitoring processes and the NIF as the only way to keep a NIF informed of the whereabouts of such long-lived processes.

Providing a reliable, built-in facility for NIFs to resolve process (or port) names simplifies these use cases considerably.

#### Risks or uncertain artifacts?

Testing has not exposed any significant risk.

The implementation behaves as expected on regular and dirty scheduler threads as well as non-scheduler threads.

By constraining the `enif_whereis_...()` functions to their minimal scopes and using patterns consistent with related functions, the implementation, testing, and maintenance burden is low.

The API and behavior of existing functions is unchanged.

#### How did you solve it?

While extending `enif_send()` to operate on a pid or an atom (as `erlang:send/2` does) was attractive, it would have entailed changing the type of its `to_pid` parameter and thereby breaking backward compatibility.

The same consideration applies to `enif_port_command()`.

That leaves a choice between 1, 2, or 3 new functions:

 1. `enif_whereis()`
 2. `enif_whereis_pid()` and `enif_whereis_port()`
 3. All of the above.

While option (1), directly mimicking the behavior of `erlang:whereis/1`, is appealing, it poses potential problems if `pid()` or `port()` are subsequently implemented as non-integral types that must be bound to an owning `ErlNifEnv` instance.

Therefore, option (2) has been chosen to use `ErlNifPid`/`ErlNifPort` structures in the API to maintain proper term ownership semantics.

sverker commented:
There is a compatibility problem with this API.

The fact that local pids and ports are immediates
is really an implementation detail. This means, enif_whereis() in its current form would need
an `out_env` argument to be future compatible. But that would be bad today as it would
mean calling `alloc_env()` just to allocate an environment that is never used.

tburghart commented:
In a previous incarnation of the code I initialized the output as
```
*out = enif_make_copy(env, res);
```
and documented it as belonging to the input environment, but changed it upon confirming that there was no way in the current implementation for the output to be anything but an integral value.

If a future implementation required it to be stored in an environment, why would it not be appropriate to store it in the one already passed into the function?

sverker commented:
The problem then is the `env` argument is used for two things; to specify the calling context and to specify where the `*out` term should be built.
Which means you can't do this from a process calling context:
```
my_env = enif_alloc_env();
enif_whereis(my_env, ...);
```

I think I would prefer `enif_whereis_pid()` with `ErlNifPid*` argument and just avoid the hassle of terms and their environments.

tburghart commented:
That does have the benefit of it being documented that `ErlNifPid` is not bound to any environment, and as the EEP lite says, it was an approach I considered.  I don't have a problem switching it to that, but it raises a couple of points.

I'm passing in the `env` because on scheduler threads the lookup wants to hold the main lock.  As the registry is self contained this doesn't seem completely necessary, but I followed that pattern as it seems to be the convention.  Unless there's a compelling reason not to, I'm inclined to continue to use the `env` in that way, but the documentation will make it clear that the resulting pid is unrelated to the environment.

Since this will be specific to pids, should there be an analogous operation for ports?  I don't see as compelling a use case for ports, but since there's an `enif_port_command` function that could use a resolved named port I'm inclined to include `enif_whereis_port` for symmetry.

The resulting API would be:
```
int enif_whereis_pid(ErlNifEnv *env, ERL_NIF_TERM name, ErlNifPid *pid);
int enif_whereis_port(ErlNifEnv *env, ERL_NIF_TERM name, ErlNifPort *port);
```

If that looks acceptable I'll make the changes in the API and tests, neither of which should pose any problem.

sverker commented:
I think the `env` argument should be documented as it is for enif_send: "The environment of the calling process. Must be NULL only if calling from a created thread."

If you see no compelling use case for `enif_whereis_port` then I will not argue for it. I wont argue against it either if it you want to do it for symmetry. But it has to be the whole shebang with docs and tests if you include it.

Another thing I thought of was for the caller to be able to differ between `name` is not an atom and `name` is not a registered process. But it's not obvious how to do that in the API and maybe it's not that important. If it fails you can always call enif_is_atom if you really need to know.

tburghart commented:
@sverker In reworking the tests to support distinct `enif_whereis_pid` and `enif_whereis_port` functions, I see there's no `enif_make_port` function to get the port from an `ErlNifPort`.

Since pid/port functions are pretty much symmetrical, is this something that should be added to the API?  Not in this PR, of course, but in a separate one?

sverker commented:
Yes, enif_make_port makes sense. But I think the principle of a real world use case is sensible to get some moderation against a growing API.

tburghart commented:
Updated to provide distinct `enif_whereis_pid()` and `enif_whereis_port()` functions.

Rebased onto `master` at ffa80a41370025ed2fb95967e731f13cc7e45e4f

@sverker I've added `enif_make_port()` as a macro in `erts/emulator/test/nif_SUITE_data/nif_SUITE.c` and `erts/emulator/test/dirty_nif_SUITE_data/dirty_nif_SUITE.c` for consistency in how the tests are written.

sverker commented:
Can you fix the return values in the documentation from `ERL_NIF_TERM` to `int`
and then squash the commit with the earlier version.
Then I'll merge it for OTP-20.0-rc2.

sverker commented:
And you need to skip all tests creating threads if `erlang:system_info(threads)` returns `false`.
Yes, it's still possible to run a beam without pthreads.
Configure with `--disable-threads` if you want to test.

tburghart commented:
Ok, will do.
I see a brand new commit on master - anything coming up that I should be sure to rebase after?

sverker commented:
No, there should be nothing that will conflict on master.

tburghart commented:
Rebased onto `master` at 7e78ec47ef00f2f7837c7a65056cf2162e0ef7e0 with requested changes.

sverker commented:
Tests crashes non-smp debug vm in `enif_send` as it's called from non-scheduler thread which is only allowed in smp enabled vm.

```
(cd $ERL_TOP/erts/emulator && make TYPE=debug plain)
$ERL_TOP/bin/cerl -debug -smp disable
```

tburghart commented:
@sverker how are you running CT so I can reproduce accurately?

sverker commented:
I do
```
(cd $ERL_TOP/lib/common_test/test_server && make release_tests)
(cd $ERL_TOP/erts/emulator/test && make release_tests)
cd $ERL_TOP/release/x86_64-unknown-linux-gnu/test_server
$ERL_TOP/bin/cerl -debug -smp disable

1> ts:install().
2> ts:run(emulator,nif_SUITE,[batch]).

```

tburghart commented:
Threaded `enif_whereis_...` tests changed to not use `enif_send`, since it has its own threaded tests.

Rebased onto `master` at dfbefdcc85b2785e4d2280c1b70de8fd14a6778d

sverker commented:
Seems we just missed the code stop for OTP-20.0-rc2, but I think it can go into 20 anyway as it's low risk on existing functionality.
